src/sinks/amqp.rs

```diff
@@ -0,0 +1,530 @@
+use crate::{
+    amqp::AmqpConfig,
+    buffers::Acker,
+    config::{log_schema, DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    event::Event,
+    internal_events::{
+        sink::{AmqpAcknowledgementFailed, AmqpDeliveryFailed, AmqpNoAcknowledgement},
+        TemplateRenderingFailed,
+    },
+    sinks::util::encoding::{EncodingConfig, EncodingConfiguration},
+    template::{Template, TemplateParseError},
+};
+use futures::{future::BoxFuture, ready, FutureExt, Sink};
+use lapin::options::BasicPublishOptions;
+use lapin::BasicProperties;
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::{
+    convert::TryFrom,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("creating amqp producer failed: {}", source))]
+    AmqpCreateFailed {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+    #[snafu(display("invalid exchange template: {}", source))]
+    ExchangeTemplate { source: TemplateParseError },
+    #[snafu(display("invalid routing key template: {}", source))]
+    RoutingKeyTemplate { source: TemplateParseError },
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct AmqpSinkConfig {
+    pub(crate) exchange: String,
+    pub(crate) routing_key: Option<String>,
+    pub(crate) encoding: EncodingConfig<Encoding>,
+    pub(crate) connection: AmqpConfig,
+}
+
+impl Default for AmqpSinkConfig {
+    fn default() -> Self {
+        Self {
+            exchange: "vector".to_string(),
+            routing_key: None,
+            encoding: Encoding::Text.into(),
+            connection: AmqpConfig::default(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Derivative, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum Encoding {
+    Text,
+    Json,
+}
+
+enum InFlight {
+    Sending(BoxFuture<'static, Result<lapin::publisher_confirm::PublisherConfirm, lapin::Error>>),
+    Committing(BoxFuture<'static, Result<lapin::publisher_confirm::Confirmation, lapin::Error>>),
+}
+
+pub struct AmqpSink {
+    channel: Arc<lapin::Channel>,
+    exchange: Template,
+    routing_key: Option<Template>,
+    encoding: EncodingConfig<Encoding>,
+    in_flight: Option<InFlight>,
+    acker: Acker,
+}
+
+inventory::submit! {
+    SinkDescription::new::<AmqpSinkConfig>("amqp")
+}
+
+impl GenerateConfig for AmqpSinkConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(
+            r#"connection_string = "amqp://localhost:5672/%2f"
+            routing_key = "user_id"
+            exchange = "test"
+            encoding.codec = "protobuf""#,
```

Comment:
```suggestion
            encoding.codec = "json""#,
```

We don't have a protobuf codec..